pkg/minishift/update/update.go

```diff
@@ -0,0 +1,305 @@
+	if err != nil {
+		return err
+	}
+
+	err = updateBinary(binary)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("\nUpdated to minishift v%s successfully.\n", latestVersion)
+
+	return nil
+}
+
+// getLatestVersionFromGitHub gets the latest version of minishift available on GitHub.
+// It returns the version and error.
+func getLatestVersionFromGitHub(githubOwner, githubRepo string) (semver.Version, error) {
+	var (
+		release *github.RepositoryRelease
+		resp    *github.Response
+		err     error
+	)
+
+	client := githubutils.Client()
+	release, resp, err = client.Repositories.GetLatestRelease(githubOwner, githubRepo)
+	if err != nil {
+		return semver.Version{}, err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	latestVersionString := release.TagName
+	if latestVersionString != nil {
+		return semver.Make(strings.TrimPrefix(*latestVersionString, "v"))
+
+	}
+
+	return semver.Version{}, fmt.Errorf("Cannot get release name.")
+}
+
+// downloadAndVerifyArchive downloads the archive of latest minishift version from GitHub
+// into a temporary location and verifies the checksum of downloaded archive.
+// It returns a string containing path to the downloaded archive.
+// It returns an error if any of the steps failed.
+func downloadAndVerifyArchive(url string) (string, error) {
+	fmt.Printf("Downloading %s\n", url)
+
+	httpResp, err := http.Get(url)
+	if err != nil {
+		return "", errors.New(fmt.Sprintf("Cannot download archive from %s: %s", url, err))
+	}
+	defer func() { _ = httpResp.Body.Close() }()
+
+	updatedArchive := httpResp.Body
+	if httpResp.ContentLength > 0 {
+		bar := pb.New64(httpResp.ContentLength).SetUnits(pb.U_BYTES)
+		bar.Start()
+		updatedArchive = bar.NewProxyReader(updatedArchive)
+		defer func() {
+			<-time.After(bar.RefreshRate)
+			fmt.Println()
+		}()
+	}
+
+	archiveBytes, err := ioutil.ReadAll(updatedArchive)
+	if err != nil {
+		return "", errors.New(fmt.Sprintf("Unable to read downloaded archive: %s", err))
+	}
+
+	// Temporary directory to store archive contents
+	dir, err := ioutil.TempDir("", "download")
+	if err != nil {
+		return "", errors.New(fmt.Sprintf("Could not create a temporary directory: %s", err))
+	}
+
+	var downloadedArchive string
+	urlSplit := strings.Split(url, "/")
+
+	downloadedArchive = filepath.Join(dir, urlSplit[len(urlSplit)-1])
+	if err := writeArchiveToDisk(downloadedArchive, archiveBytes); err != nil {
+		return "", err
+	}
+
+	checksumURL := fmt.Sprintf(url + ".sha256")
+	// Download checksum
+	checksumResp, err := http.Get(checksumURL)
+	if err != nil {
+		return "", err
+	}
+
+	defer func() { _ = checksumResp.Body.Close() }()
+
+	checksum := checksumResp.Body
+	if checksumResp.ContentLength > 0 {
+		// Newline is to separate the two progress bars
+		fmt.Println()
+		bar := pb.New64(checksumResp.ContentLength).SetUnits(pb.U_BYTES)
+		bar.Start()
+		checksum = bar.NewProxyReader(checksum)
+		defer func() {
+			<-time.After(2 * bar.RefreshRate)
+		}()
+	}
+
+	// Verify checksum
```

Comment:
I still think the checksum verification could be done much simpler with a TeeReader - see https://github.com/minishift/minishift/blob/d288208583282074afa56efe5a66004957d649b0/pkg/util/github/github.go#L150-L150